### PR TITLE
Split out package runtime reconciler from revision controller

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

Package runtimes first proposed in #4322 and implemented with #4744. However, we never did the necessary changes that would allow running external package runtimes properly. 

This PR splits package runtime logic from the revision reconciler as a separate reconciler and makes necessary refactoring and improvements along the way.  

- Splits out package deployment runtime deployment from revision reconciler, as a separate reconciler.
- Package runtime reconciler owns the `RuntimeHealthy` condition whereas revision reconciler owns `RevisionHealthy`. 
- Breaks dependency to package meta from runtime, by also resolving #6487.
- Takes care of the provider deployment selector migration - which was left as todo previously. Basically, we now use provider package name instead of provider name in meta as the label value for the key `"pkg.crossplane.io/provider"`. This was already the case for functions since the beginning.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [x] Added or updated e2e tests.
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~
- [ ] ~Followed the [API promotion workflow] if this PR introduces, removes, or promotes an API.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet
[API promotion workflow]: https://github.com/crossplane/crossplane/blob/main/contributing/guide-api-promotion.md

### CODE REVIEW

File: internal/controller/pkg/revision/reconciler.go [hunk 1]
```diff
@@ -566,7 +520,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	rewriteConfigName, newPath, err := r.config.RewritePath(ctx, imagePath)
 	if err != nil {
 		err = errors.Wrap(err, errRewriteImage)
-		pr.SetConditions(v1.Unpacking().WithMessage(err.Error()))
+		// TODO(turkenh): This was the only condition update that uses
+		//  SetConditions instead of MarkConditions. Was this intentional?
+		//  Check with @n3wscott and remove uncommented line below.
+		// pr.SetConditions(v1.Unpacking().WithMessage(err.Error()))
+		status.MarkConditions(v1.Establishing().WithMessage(err.Error()))
```
Comment (negz) [inline internal/controller/pkg/revision/reconciler.go]: CC @n3wscott 

It's also possible this got mixed up when I was rebasing the v2 branch on main.

File: internal/controller/pkg/revision/reconciler.go [hunk 2]
```diff
@@ -728,7 +663,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	// an error.
 	if rc == nil && pullPolicyNever {
 		err := errors.New(errPullPolicyNever)
-		status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
+		status.MarkConditions(v1.Establishing().WithMessage(err.Error()))
```
Comment (negz) [inline internal/controller/pkg/revision/reconciler.go]: Does this mean a revision could transition back from established to establishing if it hits an error? If so, do the errors that trigger this actually indicate the package is no longer established?

File: internal/controller/pkg/runtime/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+		err = errors.Wrap(err, errPreHook)
+		status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
+		_ = r.client.Status().Update(ctx, pr)
+		r.record.Event(pr, event.Warning(reasonSync, err))
+		return reconcile.Result{}, err
+	}
+
+	// Wait for the package revision to be established before running the
+	// post-establish hooks.
+	if pr.GetCondition(v1.TypeInstalled).Status != corev1.ConditionTrue {
```
Comment (negz) [inline internal/controller/pkg/runtime/reconciler.go]: It keeps tripping me up that this controller owns "healthy" while the main revision controller owns "installed".

I get why you did it when I think about it, but intuitively I expected:

* `Healthy` to represent the overall health of the package - so I expect the revision controller to own that
* `Installed` to represent whether the runtime was installed successfully

I'm fine leaving it as is, but wanted to point out that it wasn't intuitive for me in case you can think of a clearer approach.

File: internal/controller/pkg/runtime/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+		}
+		return reconcile.Result{Requeue: false}, nil
+	}
+
+	// Migrate provider deployment selector, if needed.
+	// Note(turkenh): This migration could be removed in the future when
+	// we no longer support older Crossplane versions that use the old
+	// provider deployment selector labels. The latest Crossplane version
+	// using the old selector labels is v1.20.0.
+	if err := r.migrateProviderDeploymentSelectors(ctx, pr, builder); err != nil {
```
Comment (negz) [inline internal/controller/pkg/runtime/reconciler.go]: I found this method name confusing. It doesn't really migrate the deployment, right? It looks like it just deleted old-style deployments so that the hooks can create a new-style one.

That said, not sure what a better name would be. `deleteLegacyDeployment` maybe?

File: internal/controller/pkg/runtime/reconciler.go [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+	return opts, nil
+}
+
+// migrateProviderDeploymentSelectors handles migration of provider deployments
+// that have outdated selector labels from older Crossplane versions.
+// The pkg.crossplane.io/provider selector format changed in newer Crossplane versions:
+// - Old format: used the provider name in the package metadata (e.g., "provider-nop")
+// - New format: uses the parent package label value (e.g., "crossplane-provider-nop", or however the provider pkg resource is named)
+// See https://github.com/crossplane/crossplane/blob/v1.20.0/internal/controller/pkg/revision/runtime_provider.go#L236-L244
+func (r *Reconciler) migrateProviderDeploymentSelectors(ctx context.Context, pr v1.PackageRevisionWithRuntime, builder ManifestBuilder) error {
```
Comment (negz) [inline internal/controller/pkg/runtime/reconciler.go]: Should this be a method on `Reconciler`?

When I find jobs like this that I want to break out of the reconciler, I prefer to make them injectable dependencies rather than methods. This way:

* I can easily inject a dummy no-op or test implementation when I'm unit testing `Reconciler` - I don't "test through" to this method when testing `Reconciler`
* If we ever identify an alternative implementation it's easy to swap in another implementation

One way I decide whether something should be an injected dependency or a method is by looking at how many fields of `Reconciler` the new function needs to access. Here it looks like the new method only uses `r.client` and `r.log`.

So I'd prefer to:

1. Add a new `DeploymentSelectorMigrator` interface (or whatever name feels appropriate)
2. Add a `WithDeploymentSelectorMigrator` `ReconcilerOption`
3. Make a new type with a single method, e.g.:

```go
type DeploymentSelectorMigrator interface {
    MigrateDeploymentSelector(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error
}

type DeletingDeploymentSelectorMigrator struct {
    c   client.Client
    log logging.Logger
}

func (m *DeletingDeploymentSelectorMigrator) MigrateDeploymentSelector(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error {
    // TODO
}
```

**Review by negz — COMMENTED** [review]
Direction looks good to me - looks like all the E2E tests are failing though.

File: internal/controller/pkg/revision/reconciler.go (hunk 2 above)
Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go]: Some background: `Establishing()` and `Established()` using the already existing `Installed` condition under the hood, I didn't introduce a new condition rather tried to use `Installed` which was only used by packages so far, not in revisions. I am happy to name this helpers to something else, but I don't think we need to introduce a new condition type as I believe `Installed` is a good fit here. 

> Does this mean a revision could transition back from established to establishing if it hits an error? If so, do the errors that trigger this actually indicate the package is no longer established? 

Yes, basically, it could go from `Installed=True` to installed `Installed=False`. For example, the error you commented happens when package is not in the cache and pull policy is never. In this case, we will communicate this with an error message on Installed condition by setting it to False. I can see your point on how this could be confusing; it doesn't mean the package is not installed at all but means installation per latest desired state is not completed successfully. FWIW, I believe we are following similar approach on other reconcilers as well, for example, if MR reconciliation fails due to an error we mark it as `Synced=False` with the error message. This doesn't mean, we never synced the MR to the external API, but this time, we couldn't complete it successfully per latest desired state. Could `Installed` have a little different implications than `Synced`? Probably, but I don't think it is a big problem.

File: internal/controller/pkg/runtime/reconciler.go (hunk 3 above)
Comment (turkenh) [reply internal/controller/pkg/runtime/reconciler.go]: Yeah, it depends on where you look at it and here is how I was seeing them:

`Installed` means, you installed the package contents on to the API. CRDs in case of providers, compositions/XRDs in case of configurations. 
`Healthy` means, your package installation is functional as expected. So, you can start using whatever your package brought to you.

> `Healthy` to represent the overall health of the package - so I expect the revision controller to own that

FWIW, even though the runtime reconciler owns `Healthy` condition, `Installed=True` is a prerequisite for being `Healthy`. So, actually the overall healthy is owned by both controllers.

File: internal/controller/pkg/runtime/reconciler.go (hunk 5 above)
Comment (turkenh) [reply internal/controller/pkg/runtime/reconciler.go]: I initially started by adding a `Migrate` function to the runtime.Hooks interface: 

```go
// A Hooks performs runtime operations before and after a revision
// establishes objects.
type Hooks interface {
	// Migrate performs migrations, if any, before starting reconciling the revision.
	Migrate(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error

	// Pre performs operations meant to happen before establishing objects.
	Pre(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error

	// Post performs operations meant to happen after establishing objects.
	Post(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error

	// Deactivate performs operations meant to happen before deactivating a revision.
	Deactivate(ctx context.Context, pr v1.PackageRevisionWithRuntime, b ManifestBuilder) error
}
```

It aligns well with the rest of other methods there. Later, I thought this would be a little overkill since we only need this now, only for providers and probably clean it up in on of the next releases (this is the only migration we needed since the inception of packages, to the best of my knowledge). This would apply what you're proposing. 

Fine to switch back to that approach if you think it is still worth it.

File: internal/controller/pkg/runtime/reconciler.go (hunk 4 above)
Comment (turkenh) [reply internal/controller/pkg/runtime/reconciler.go]: Fine with that name, but I think we first need to align with the other comment below.

**Comment by turkenh** [comment]
> Direction looks good to me - looks like all the E2E tests are failing though.

Found a couple of issues and all should be fixed right now 🤞 Great to see that 95% of e2e failures were real issues, not flakes or test case related problems.

Depending on your responses to some open threads, I feel I am close to completion.

File: internal/controller/pkg/runtime/reconciler.go (hunk 5 above)
Comment (negz) [reply internal/controller/pkg/runtime/reconciler.go]: That seems reasonable to me. I think I prefer it to the method.

Given we think migration will be an edge case you could also:

1. Make `Migrator` its own interface, outside of hooks.
2. Have types that satisfy the `Hooks` interface also optionally satisfy the `Migrator` interface
3. Test whether the `Hook` implementation also satisfies `Migrator`, and only call it if it does

File: internal/controller/pkg/runtime/reconciler.go (hunk 3 above)
Comment (negz) [reply internal/controller/pkg/runtime/reconciler.go]: Thinking about this a little more with https://github.com/crossplane/crossplane/pull/6498#discussion_r2123583502 also in mind.

Currently in main the `Installed` condition has three reasons - `Unpacking`, `Inactive`, and `Active`. When the controller hits an error it sets the `Healthy` condition to false. So in main `Installed` mostly tracks whether the package is active or inactive, while `Healthy` represents whether the package is "working" - e.g. whether it's reconciling without issues.

This PR changes the meaning of `Installed` and `Healthy`:

* Now `Healthy` represents whether the runtime is working - e.g. whether it's reconciling without issues.
* Now `Installed` represents both whether the package is active and whether it's healthy.

Instead of changing/extending the meaning of `Installed`, could we break `Healthy` into subconditions? For example:

* `Installed` - Meaning stays the same - represents whether the package is unpacking, active, or inactive
* `PackageHealthy` - Represents whether the package was pulled, unpacked, dependencies resolved, etc
* `RuntimeHealthy` - Represents whether a runtime was started for the package
* `Healthy` - Represents basically `PackageHealthy && RuntimeHealthy`

This way the existing meaning of `Healthy` and `Installed` don't change, but each controller owns its own set of conditions.

File: internal/controller/pkg/revision/reconciler.go (hunk 1 above)
Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go]: I removed this todo assuming it is fine.

File: internal/controller/pkg/runtime/reconciler.go (hunk 3 above)
Comment (turkenh) [reply internal/controller/pkg/runtime/reconciler.go]: Makes sense. 
Updated accordingly, except used `RevisionHealthy` instead of `PackageHealthy` since managed by revision controller.

File: internal/controller/pkg/runtime/reconciler.go (hunk 5 above)
Comment (turkenh) [reply internal/controller/pkg/runtime/reconciler.go]: Updated per your first comment, didn't see any benefit with implementing this in runtime hooks.

File: internal/controller/pkg/manager/reconciler.go [hunk 6]
```diff
@@ -564,6 +551,38 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	return pullBasedRequeue(p.GetPackagePullPolicy()), errors.Wrap(r.client.Status().Update(ctx, p), errUpdateStatus)
 }
 
+func packageHealth(pr v1.PackageRevision) xpv1.Condition {
```
Comment (negz) [inline internal/controller/pkg/manager/reconciler.go]: WDYT about making this a helper in the same file that defines the conditions?

File: internal/controller/pkg/manager/reconciler.go [hunk 7]
```diff
@@ -564,6 +551,38 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+	if revisionHealthy && runtimeHealthy {
+		return v1.Healthy()
+	}
+
+	if !revisionHealthy {
+		m := fmt.Sprintf("Package revision health is %q", revisionHealth.Status)
+		if revisionHealth.Message != "" {
+			m += " with message: " + revisionHealth.Message
+		}
+		return v1.Unhealthy().WithMessage(m)
```
Comment (negz) [inline internal/controller/pkg/manager/reconciler.go]: Why add this message? If the user's reading this status condition message, won't they also see the `RevisionHealthy` status condition message right next to it? Seems a bit redundant to me.

File: internal/controller/pkg/revision/reconciler.go [hunk 8]
```diff
@@ -678,12 +609,12 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 			// status.objectRefs is empty, to make sure that the revision is
 			// removed from the lock which could otherwise block a successful
 			// reconciliation.
-			if pr.GetCondition(v1.TypeHealthy).Status != corev1.ConditionTrue {
+			if pr.GetCondition(v1.TypeInstalled).Status != corev1.ConditionTrue {
```
Comment (negz) [inline internal/controller/pkg/revision/reconciler.go]: Should this still be `TypeInstalled`?

File: test/e2e/install_test.go [hunk 9]
```diff
@@ -200,7 +223,7 @@ func TestCrossplaneLifecycle(t *testing.T) {
 				funcs.DeleteResources(manifests, "claim.yaml"),
 				funcs.ResourcesDeletedWithin(2*time.Minute, manifests, "claim.yaml"),
 			)).
-			WithTeardown("DeletePrerequisites", funcs.ResourcesDeletedAfterListedAreGone(3*time.Minute, manifests, "setup/*.yaml", nopList)).
+			// WithTeardown("DeletePrerequisites", funcs.ResourcesDeletedAfterListedAreGone(3*time.Minute, manifests, "setup/*.yaml", nopList)).
```
Comment (negz) [inline test/e2e/install_test.go]: This should be uncommented or deleted. 🙂

**Review by negz — APPROVED** [review]
Overall this LGTM so approving to unblock.

Please take a look at my last batch of comments though - one or two of them might be real things that need fixing.

File: internal/controller/pkg/manager/reconciler.go (hunk 7 above)
Comment (turkenh) [reply internal/controller/pkg/manager/reconciler.go]: Not really. 
`RevisionHealthy` is a status condition (similar to `RuntimeHealthy`) on Package Revisions not on the Package object.

File: internal/controller/pkg/revision/reconciler.go (hunk 8 above)
Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go]: Nope, good catch!

File: apis/pkg/v1/function_types.go [hunk 10]
```diff
@@ -85,13 +85,14 @@ type FunctionRevisionSpec struct {
...
+// +kubebuilder:printcolumn:name="HEALTHY",type="string",JSONPath=".status.conditions[?(@.type=='RevisionHealthy')].status"
+// +kubebuilder:printcolumn:name="RUNTIME",type="string",JSONPath=".status.conditions[?(@.type=='RuntimeHealthy')].status"
 // +kubebuilder:printcolumn:name="IMAGE",type="string",JSONPath=".spec.image"
 // +kubebuilder:printcolumn:name="STATE",type="string",JSONPath=".spec.desiredState"
-// +kubebuilder:printcolumn:name="DEP-FOUND",type="string",JSONPath=".status.foundDependencies"
-// +kubebuilder:printcolumn:name="DEP-INSTALLED",type="string",JSONPath=".status.installedDependencies"
 // +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:printcolumn:name="REVISION",type="string",JSONPath=".spec.revision",priority=1
+// +kubebuilder:printcolumn:name="DEP-FOUND",type="string",JSONPath=".status.foundDependencies",priority=1
+// +kubebuilder:printcolumn:name="DEP-INSTALLED",type="string",JSONPath=".status.installedDependencies",priority=1
```
Comment (turkenh) [inline apis/pkg/v1/function_types.go:95]: Moved these to `-o wide` as we have many columns and these are relatively less needed compared to the others.

File: internal/controller/pkg/revision/reconciler.go [hunk 11]
```diff
@@ -615,40 +563,39 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
...
+		// Update the applied image config refs and persist immediately
+		pr.ClearAppliedImageConfigRef(v1.ImageConfigReasonSetPullSecret)
+		if psRef != nil {
+			log.Debug("Selected pull secret from image config store", "image", pr.GetResolvedSource(), "imageConfig", pullSecretConfig, "pullSecret", pullSecretFromConfig, "rewriteConfig", rewriteConfigName)
+			r.record.Event(pr, event.Normal(reasonImageConfig, fmt.Sprintf("Selected pullSecret %q from ImageConfig %q for registry authentication", pullSecretFromConfig, pullSecretConfig)))
+			pr.SetAppliedImageConfigRefs(*psRef)
 		}
-
-		runtimeManifestBuilder = NewRuntimeManifestBuilder(pwr, r.namespace, opts...)
+		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
```
Comment (turkenh) [inline internal/controller/pkg/revision/reconciler.go:593]: We need to ensure that applied image config for pull secret is persisted to the status since runtime reconciler relies on this instead of resolving in both places. 

Previously, it was not persisted since there was a client.Update before a status update.

**Comment by turkenh** [comment]
@negz I did some more testing, identified a couple of more issues and fixed them in the last 3 commits. Left comments for the ones that deserves highlighting. Please let me know if you want to have another look or whether your approval is still valid.

File: internal/controller/pkg/manager/reconciler.go (hunk 7 above)
Comment (negz) [reply internal/controller/pkg/manager/reconciler.go]: Oh got it - I missed that this was setting a condition on the package, not the revision.
